src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs

```diff
@@ -104,11 +110,201 @@ public int Execute(string moduleName, string[] args, string expectedOutput)
+            {
+                if (imagesByName.TryGetValue(name, out var image))
+                {
+                    return new PEReader(image);
+                }
+
+                return null;
+            }
+        }
+
         public void Verify(Verification verification)
         {
             var emitData = GetEmitData();
             emitData.RuntimeData.PeverifyRequested = true;
             // TODO(https://github.com/dotnet/coreclr/issues/295): Implement peverify
+
+            // IL Verify
+            if (verification == Verification.Skipped)
+            {
+                return;
+            }
+
+            try
+            {
+                var resolver = new Resolver(emitData);
+                var verifier = new ILVerify.Verifier(resolver);
+                var mscorlibModules = emitData.AllModuleData.Where(m => m.SimpleName == "mscorlib").ToArray();
+                if (mscorlibModules.Length == 1)
+                {
+                    verifier.SetSystemModuleName(new AssemblyName(mscorlibModules[0].FullName));
+                }
+                else
+                {
+                    // auto-detect which module is the "corlib"
+                    foreach (var module in emitData.AllModuleData)
+                    {
+                        var name = module.SimpleName;
+                        var metadataReader = resolver.Resolve(name).GetMetadataReader();
+                        if (metadataReader.AssemblyReferences.Count == 0)
+                        {
+                            try
+                            {
+                                verifier.SetSystemModuleName(new AssemblyName(name));
+                            }
+                            catch (Exception ex)
+                            {
+                                // ILVerify checks that corlib contains certain types
+                                if ((verification & Verification.FailsIlVerify_MissingStringType) == Verification.FailsIlVerify_MissingStringType
+                                    && ex.Message.Contains("Failed to load type 'System.String' from assembly"))
+                                {
+                                    return;
+                                }
+
+                                throw;
+                            }
+                        }
+                    }
+                }
+
+                var result = verifier.Verify(resolver.Resolve(emitData.MainModule.FullName));
+                if (result.Count() == 0)
+                {
+                    if ((verification & Verification.FailsIlVerify) == 0)
+                    {
+                        return;
+                    }
+
+                    throw new Exception("IL Verify succeeded unexpectedly");
+                }
+
+                string message = printVerificationResult(result);
+                if ((verification & Verification.FailsIlVerify_TypedReference) == Verification.FailsIlVerify_TypedReference
```

Comment:
Curious: our previous verifaction was mostly yes / no where this implementation is more looking for the exact failure case. Is there a reason you chose this route? I worry a bit that we're using messages to infer error codes. It would seem more natural to push strongly typed error codes into ILVerify if that was our desire here.

Reply:
The main reason I thought it useful to break down the kinds of failures is that some of those are due to limitations in ILVerify or to the verification rules. Having a breakdown of the failures would help us inform where to improve ILVerify.
If you think we should stick with pass vs. skip, that's good feedback since it will much simplify the PR

Reply:
For the initial PR I would favor sticking with our existing pass / fail semantics. If we have a desire to make it more complicated I think that's reasonable but I would start by pushing ILVerify to produce diagnostics that are more strongly typed (an `enum`, error code, etc ...).